### PR TITLE
Sanketika-Obsrv/issue-tracker#68: feat: pseudocode for the dataset update api

### PR DESCRIPTION
**Pull request by JeraldJF**

### CODE REVIEW

File: api-service/src/configs/DatasetConfigDefault.ts [hunk 1]
```diff
@@ -32,12 +32,11 @@ export const defaultMasterConfig = {
     "dataset_config": {
         "data_key": "",
         "timestamp_key": "",
-        "exclude_fields": [],
```
Comment (HarishGangula) [inline api-service/src/configs/DatasetConfigDefault.ts:35]: Check  with manju

File: api-service/src/controllers/DatasetUpdate/DatasetUpdateValidationSchema.json [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+      "minLength": 1
+    },
+    "validation_config": {
+      "type": "object",
+      "properties": {
+        "validate": {
+          "type": "boolean"
+        },
+        "mode": {
+          "type": "string",
```
Comment (HarishGangula) [inline api-service/src/controllers/DatasetUpdate/DatasetUpdateValidationSchema.json:20]: This should be enum

File: api-service/src/controllers/DatasetUpdate/DatasetUpdateValidationSchema.json [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+        "properties",
+        "additionalProperties"
+      ],
+      "additionalProperties": false
+    },
+    "dataset_config": {
+      "type": "object",
+      "properties": {
+        "data_key": {
+          "type": "string"
```
Comment (HarishGangula) [inline api-service/src/controllers/DatasetUpdate/DatasetUpdateValidationSchema.json:130]: minLength should be added

File: api-service/src/controllers/DatasetUpdate/DatasetUpdateValidationSchema.json [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+                "type": "string",
+                "minLength": 1
+              },
+              "transformation_function": {
+                "type": "string",
+                "minLength": 1
+              },
+              "mode": {
+                "type": "string",
+                "minLength": 1
```
Comment (HarishGangula) [inline api-service/src/controllers/DatasetUpdate/DatasetUpdateValidationSchema.json:157]: Add the enum for this also

**Review by HarishGangula — CHANGES_REQUESTED** [review]

File: api-service/src/configs/DatasetConfigDefault.ts (hunk 1 above)
Comment (JeraldJF) [reply api-service/src/configs/DatasetConfigDefault.ts:35]: Confirmed with manju, exclude fields parameter not required to default

**Comment by JeraldJF** [comment]
API development done, PR review is pending along with test cases

File: api-service/src/controllers/DatasetCreate/DatasetCreate.ts [hunk 5]
```diff
@@ -25,7 +26,8 @@ const datasetCreate = async (req: Request, res: Response) => {
             } as ErrorObject, req, res);
         }
 
-        const isDataSetExists = await checkDatasetExists(_.get(req, ["body", "dataset_id"]));
+        const datasetId = id || `${dataset_id}.1`
```
Comment (HarishGangula) [inline api-service/src/controllers/DatasetCreate/DatasetCreate.ts:29]: Don't include creation code in update

File: api-service/src/controllers/DatasetUpdate/DatasetUpdate.ts [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+                message: isRequestValid.message,
+                statusCode: 400,
+                errCode: "BAD_REQUEST"
+            } as ErrorObject, req, res);
+        }
+
+        const { dataset_id, ...rest } = datasetBody
+        if (_.isEmpty(rest)) {
+            return ResponseHandler.errorResponse({
+                message: "Invalid request body",
```
Comment (HarishGangula) [inline api-service/src/controllers/DatasetUpdate/DatasetUpdate.ts:30]: Message should be proper says what is not correct
`Atleast one field is required to updated dataset other than dataset_id`

File: api-service/src/controllers/DatasetUpdate/DatasetUpdate.ts [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+            return ResponseHandler.errorResponse({
+                message: "Dataset does not exists",
+                statusCode: 404,
+                errCode: "NOT_FOUND"
+            } as ErrorObject, req, res);
+        }
+
+        if (isDatasetExists && datasetStatus != DatasetStatus.Draft) {
+            return ResponseHandler.errorResponse({
+                message: "Given record is not a draft dataset",
```
Comment (HarishGangula) [inline api-service/src/controllers/DatasetUpdate/DatasetUpdate.ts:47]: `Dataset is not in draft state to update`

File: api-service/src/controllers/DatasetUpdate/DatasetUpdate.ts [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+                errCode: "NOT_FOUND"
+            } as ErrorObject, req, res);
+        }
+
+        const duplicateDenormKeys = getDuplicateDenormKey(_.get(datasetBody, "denorm_config"))
+        if (!_.isEmpty(duplicateDenormKeys)) {
+            logger.error(`Duplicate denorm output fields found. Duplicate Denorm out fields are [${duplicateDenormKeys}]`)
+            return ResponseHandler.errorResponse({
+                statusCode: 400,
+                message: "Duplicate denorm output fields found",
```
Comment (HarishGangula) [inline api-service/src/controllers/DatasetUpdate/DatasetUpdate.ts:58]: Add what fields are duplicate in message

File: api-service/src/controllers/DatasetUpdate/DatasetUpdate.ts [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+
+const setDedupConfigs = (configs: Record<string, any>): Record<string, any> => {
+    const dropDuplicates = _.get(configs, "drop_duplicates")
+    if (!dropDuplicates) {
+        return { ...defaultDatasetConfig.dedup_config, drop_duplicates: dropDuplicates }
+    }
+    return configs
+}
+
+const getUpdatedTags = (payload: Record<string, any>, datasetTags: Array<string>): Record<string, any> => {
```
Comment (HarishGangula) [inline api-service/src/controllers/DatasetUpdate/DatasetUpdate.ts:153]: Error scenarios
` action === "add" || "remove"
ptags = lower(pTags)
 pTags = uniq(pTags)

1. Action == "add" && includes()
2. Action ==. "remove" && !included()

1. action === "add" {
   tags = _.lower([..existing , ...provided])
  // update the tags in table
}
1. action === "remove" {
   tags = _.lower(filter(tags, pTags))
  // update the tags in table
}`

**Review by HarishGangula — CHANGES_REQUESTED** [review]

**Review by HarishGangula — APPROVED** [review]

**Pull request merged**